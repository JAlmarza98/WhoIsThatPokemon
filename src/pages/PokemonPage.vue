<template>
  <h1 v-if="!pokemon">Wait a moment please...</h1>
  <div v-else>
    <h1>who's that pokemon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer"/>

    <div v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">
        New game
      </button>
    </div>

    <div class="score">
      <div>score</div>
      <div>{{score}}</div>
    </div>

  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'

import getPokemonOptions from '../helpers/getPokemonsOptions'

export default {
  components: {
    PokemonPicture,
    PokemonOptions
  },
  data () {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: '',
      score: 0
    }
  },
  methods: {
    async mixPokemonArray () {
      this.pokemonArr = await getPokemonOptions()

      const rndInt = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonArr[rndInt]
    },
    checkAnswer (selectedId) {
      this.showPokemon = true
      this.showAnswer = true

      if (selectedId === this.pokemon.id) {
        this.message = `Yes, ${this.pokemon.name}`
        this.score++
      } else {
        this.message = `Oops, it was ${this.pokemon.name}`
        this.score = 0
      }
    },
    newGame () {
      this.showPokemon = false
      this.showAnswer = false
      this.pokemon = null
      this.pokemonArr = []
      this.mixPokemonArray()
    }
  },
  mounted () {
    this.mixPokemonArray()
  }
}
</script>

<style>
  h1,h2{
    color: white;
  }
  button {
    background-color: white;
    padding: 10px 0 10px 0;
    border-radius: 5px;
    border: 1px solid rgba(0, 0, 0, 0.2);
    cursor: pointer;
    margin-bottom: 10px;
    width: 250px;
    font-size: larger;
  }
  button:hover{
    background-color: rgb(192, 192, 192);
  }
  .score {
    background-color: white;
    border: 1px solid rgba(0, 0, 0, 0.2);
    padding: 15px;
    width: fit-content;
    margin: 0 auto;
    border-radius: 8px;
    font-size: larger;
    font-weight: bold;
  }
</style>
