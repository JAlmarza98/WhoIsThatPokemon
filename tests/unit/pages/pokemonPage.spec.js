import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage.vue'

import { mockPokemons } from '../../mocks/pokemons.mock'

describe('PokemonOptions Component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage)
  })

  test('Debe hacer match con el Snapshot sin datos', () => {
    expect( wrapper.html() ).toMatchSnapshot()
  })

  test('Debe hacer match con el Snapshot con datos', () => {
    const wrapper = mount(PokemonPage, {
      data () {
        return {
          pokemonArr: mockPokemons,
          pokemon: mockPokemons[1],
          showPokemon: false,
          showAnswer: false,
          message: '',
          score: 0
        }
      }
    })

    expect( wrapper.html() ).toMatchSnapshot()
  })

  test('Debe de llamar mixPokemonArray al montarse', () => {
    const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray')
    shallowMount(PokemonPage)

    expect(mixPokemonArraySpy).toHaveBeenCalled()
  })

  test('Debe mostar los compoenentes PokemonPicture y PokemonOptions', () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemonArr: mockPokemons,
          pokemon: mockPokemons[1],
          showPokemon: false,
          showAnswer: false,
          message: '',
          score: 0
        }
      }
    })

    const picture = wrapper.find('pokemon-picture-stub')
    const options = wrapper.find('pokemon-options-stub')

    expect(picture.exists()).toBeTruthy()
    expect(options.exists()).toBeTruthy()

    expect(picture.attributes('pokemonid')).toBe(`${mockPokemons[1].id}`)
    expect(options.attributes('pokemons')).toBeTruthy()
  })

  test('Puebas en checkAnswer', async() => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemonArr: mockPokemons,
          pokemon: mockPokemons[1],
          showPokemon: false,
          showAnswer: false,
          message: '',
          score: 0
        }
      }
    })

    await wrapper.vm.checkAnswer(mockPokemons[1].id)
    expect(wrapper.find('h2').exists()).toBeTruthy()
    expect(wrapper.find('h2').text()).toBe(`Yes, ${mockPokemons[1].name}`)
    expect(wrapper.vm.showPokemon).toBeTruthy()

    await wrapper.vm.checkAnswer(mockPokemons[2].id)
    expect(wrapper.vm.message).toBe(`Oops, it was ${mockPokemons[1].name}`)

  })

})
