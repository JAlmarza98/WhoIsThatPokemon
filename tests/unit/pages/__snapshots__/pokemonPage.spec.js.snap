// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonOptions Component Debe hacer match con el Snapshot con datos 1`] = `
<div>
  <h1>who's that pokemon?</h1>
  <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/2.svg" class="hidden-pokemon" alt="pokemon"></div>
  <div class="options-container">
    <ul>
      <li>bulbasaur</li>
      <li>ivysaur</li>
      <li>venusaur</li>
      <li>charmander</li>
    </ul>
  </div>
  <!--v-if-->
  <div class="score">
    <div>score</div>
    <div>0</div>
  </div>
</div>
`;

exports[`PokemonOptions Component Debe hacer match con el Snapshot sin datos 1`] = `<h1>Wait a moment please...</h1>`;
